timelock-impl/src/main/java/com/palantir/atlasdb/timelock/ConjureTimelockResource.java

```diff
@@ -0,0 +1,149 @@
+import com.palantir.conjure.java.undertow.lib.UndertowService;
+import com.palantir.leader.NotCurrentLeaderException;
+import com.palantir.lock.impl.TooManyRequestsException;
+import com.palantir.lock.remoting.BlockingTimeoutException;
+import com.palantir.lock.v2.ImmutableStartTransactionRequestV5;
+import com.palantir.lock.v2.LeaderTime;
+import com.palantir.lock.v2.StartTransactionRequestV5;
+import com.palantir.lock.v2.StartTransactionResponseV5;
+import com.palantir.tokens.auth.AuthHeader;
+
+public final class ConjureTimelockResource implements UndertowConjureTimelockService {
+    private final RedirectRetryTargeter redirectRetryTargeter;
+    private final Function<String, AsyncTimelockService> timelockServices;
+
+    @VisibleForTesting
+    ConjureTimelockResource(
+            RedirectRetryTargeter redirectRetryTargeter,
+            Function<String, AsyncTimelockService> timelockServices) {
+        this.redirectRetryTargeter = redirectRetryTargeter;
+        this.timelockServices = timelockServices;
+    }
+
+    public static UndertowService undertow(
+            RedirectRetryTargeter redirectRetryTargeter,
+            Function<String, AsyncTimelockService> timelockServices) {
+        return ConjureTimelockServiceEndpoints.of(new ConjureTimelockResource(redirectRetryTargeter, timelockServices));
+    }
+
+    public static ConjureTimelockService jersey(
+            RedirectRetryTargeter redirectRetryTargeter,
+            Function<String, AsyncTimelockService> timelockServices) {
+        return new JerseyAdapter(new ConjureTimelockResource(redirectRetryTargeter, timelockServices));
+    }
+
+    @Override
+    public ListenableFuture<ConjureStartTransactionsResponse> startTransactions(
+            AuthHeader authHeader, String namespace, ConjureStartTransactionsRequest request) {
+        return handleExceptions(() -> {
+            StartTransactionRequestV5 legacyRequest = ImmutableStartTransactionRequestV5.builder()
+                    .requestId(request.getRequestId())
+                    .requestorId(request.getRequestorId())
+                    .numTransactions(request.getNumTransactions())
+                    .build();
+            StartTransactionResponseV5 response = tl(namespace).startTransactionsWithWatches(legacyRequest);
+            return ConjureStartTransactionsResponse.builder()
+                    .immutableTimestamp(response.immutableTimestamp())
+                    .timestamps(response.timestamps())
+                    .lease(response.lease())
+                    .build();
+        });
+    }
+
+    @Override
+    public ListenableFuture<LeaderTime> leaderTime(AuthHeader authHeader, String namespace) {
+        return handleExceptions(() -> tl(namespace).leaderTime());
+    }
+
+    private AsyncTimelockService tl(String namespace) {
+        return timelockServices.apply(namespace);
+    }
+
+    private <T> ListenableFuture<T> handleExceptions(Supplier<T> supplier) {
```

Comment:
let's have the `Supplier<T>` be a `Callable<T>` which is a bit more accurate.

Reply:
I would rather not callers be able to use checked exceptions.